src/lib/Builder.hs

```diff
@@ -70,41 +73,35 @@ type SubstBuilderT m = ReaderT SubstEnv (BuilderT m)
 type SubstBuilder    = SubstBuilderT Identity
 
 -- Carries the vars in scope (with optional definitions) and the emitted decls
-type BuilderEnvC = (Scope, Nest Decl)
+-- TODO: The only reason to have the inference candidates here is that we want
+-- to use traversal template in dictinoary synthesis. We should come up with a
+-- different design that's more modular.
```

Comment:
Not easily, because of the way we use the `TraversalDef` traversal from Builder during synthesis, which doesn't let you parameterize how to go under lambda binders.

Reply:
I'm not sure if I follow. `TraversalDef` does allow you to use other monads and you could override `traverseAtom` to decide what happens when you go under a lambda

Reply:
Ah, good point! I hadn't thought of that. `TraversalDef` is more expressive than I'd realized.

Reply:
Just tried it. It doesn't quite work because we also need to update the synthesis candidates for local let-bound instances, but the decl traversal looks like `Decl -> m SubstEnv`, whereas we'd need it to be something like `Decl -> m a -> m a`. I'm going to leave it as-is for now, because the safer-names system will change it anyway.